solr/core/src/java/org/apache/solr/response/transform/DeeplyNestedChildDocTransformer.java

```diff
@@ -0,0 +1,214 @@
+      final int seg = (int) (segAndId >> 32);
+      final LeafReaderContext leafReaderContext = searcher.getIndexReader().leaves().get(seg);
+      final SortedDocValues segPathDocValues = DocValues.getSorted(leafReaderContext.reader(), NEST_PATH_FIELD_NAME);
+
+      Multimap<String,SolrDocument> pendingParentPathsToChildren = ArrayListMultimap.create();
+
+      if(children.matches() > 0) {
+        SolrDocumentFetcher docFetcher = searcher.getDocFetcher();
+        Set<String> dvFieldsToReturn = docFetcher.getNonStoredDVs(true).stream()
+            .filter(name -> !NEST_PATH_FIELD_NAME.equals(name)).collect(Collectors.toSet());
+        boolean shouldDecorateWithDVs = dvFieldsToReturn.size() > 0;
+        DocIterator i = children.iterator();
+        final int segBaseId = leafReaderContext.docBase;
+        final int firstChildDocId = i.nextDoc();
+        assert firstChildDocId < rootDocId;
+
+        for (int docId = firstChildDocId; docId < rootDocId; ++docId) {
+          // get the path
+          String fullDocPath = getPathByDocId(docId - segBaseId, segPathDocValues);
+
+          // Is this doc a direct ancestor of another doc we've seen?
+          boolean isAncestor = pendingParentPathsToChildren.containsKey(fullDocPath);
+
+          // Do we need to do anything with this doc (either ancestor or a matched the child query)
+          if (isAncestor || children.exists(docId)) {
+            // load the doc
+            SolrDocument doc = DocsStreamer.convertLuceneDocToSolrDoc(docFetcher.doc(docId),
+                schema, new SolrReturnFields());
+            doc.setField(NEST_PATH_FIELD_NAME, fullDocPath);
+            if (shouldDecorateWithDVs) {
+              docFetcher.decorateDocValueFields(doc, docId, dvFieldsToReturn);
+            }
+            // get parent path
+            // put into pending
+            String parentDocPath = lookupParentPath(fullDocPath);
+            pendingParentPathsToChildren.put(parentDocPath, doc); // multimap add (won't replace)
+
+            // if this path has pending child docs, add them.
+            if (isAncestor) {
+              addChildrenToParent(doc, pendingParentPathsToChildren.get(fullDocPath));
+              pendingParentPathsToChildren.removeAll(fullDocPath); // no longer pending
+            }
+          }
+        }
+
+        // only children of parent remain
+        assert pendingParentPathsToChildren.keySet().size() == 1;
+
+        addChildrenToParent(rootDoc, pendingParentPathsToChildren.get(null));
+      }
+    } catch (IOException e) {
+      rootDoc.put(getName(), "Could not fetch child Documents");
+    }
+  }
+
+  void addChildToParent(SolrDocument parent, SolrDocument child, String label) {
+    // lookup leaf key for these children using path
+    // depending on the label, add to the parent at the right key/label
+    // TODO: unfortunately this is the 2nd time we grab the paths for these docs. resolve how?
```

Comment:
I think this TODO is not an issue based on your placement of the path in the doc.